synapse/rest/media/v1/media_storage.py

```diff
@@ -302,3 +316,44 @@ def write_to_consumer(self, consumer: IConsumer) -> Deferred:
 
     def __exit__(self, exc_type, exc_val, exc_tb):
         self.open_file.close()
+
+
+class SpamMediaException(NotFoundError):
+    """The media was blocked by a spam checker, so we simply 404 the request (in
+    the same way as if it was quarantined).
+    """
+
+    pass
+
+
+@attr.s(slots=True)
+class ReadableFileWrapper:
+    """Wrapper that allows reading a file in chunks, yielding to the reactor,
+    and writing to an IO stream.
+
+    This is simplified `FileSender` that takes an IO object rather than an
+    `IConsumer`.
+    """
```

Comment:
The rationale here is that we don't want to give the raw path to the spam checker so that people don't just do `file.read()`, potentially blocking the reactor while we do large reads. Instead we do what `FileSender` does and read small chunks, yielding to the reactor between reads. I'm not 100% if that's the right way of doing things, but at least it's what we currently do for file downloads.

Reply:
it's really not the right way of doing things. Irrespective of whether non-blocking IO is perfect, it's better than doing regular IO on the reactor thread.

It's probably good enough for a quick hack, but it would be nice to come back to it and either make a non-blocking-io thing which integrates with the twisted reactor, or give up and uses threads.

Reply:
Have filed https://github.com/matrix-org/synapse/issues/9312